doc/source/f2py/usage.rst

```diff
@@ -86,6 +86,13 @@ Here ``<fortran files>`` may also contain signature files. Among other options
   ``--wrap-functions`` is default because it ensures maximum portability and
   compiler independence.
 
+``--[no-]requires-gil``
+  Create a module that declares it does or doesn't require the GIL. The default
+  is ``--requires-gil`` for backwards compatibility. Inspect the fortran
+  code you are wrapping for thread safety issues before passing
+  ``--no-requires-gil``, as ``f2py`` does not analyze fortran code for thread
+  safety issues.
```

Comment:
There were two things that came up on the other PR.  One was the idea to align with cython and use `--freethreading-compatible`, not sure about the merit, but as such aligning sounds like a good idea.

The other was that existing `threadsafe` markers are already sufficient to mark a module as compatible if all functions are marked.
So unless you actually don't like GIL release (due to overhead?), in all cases where you pass `--no-requires-gil` it makes sense to also mark all functions as `threadsafe` for the GIL enabled builds.
(If you do pass `--requires-gil` you will get a `RuntimeWarning` either way, I think; the downside being that I suppose that runtime warning will tell you the module, but less about how to adapt.)

Reply:
I looked at this today and it's a little more complicated than just assuming that if the GIL is disabled, we can mark functions as thread safe. In particular, the handling for callbacks uses the `PyThreadState` C API, which cannot be accessed inside `Py_BEGIN_ALLOW_THREADS` blocks.

My understanding is that in the free-threaded build, the adding the `Py_BEGIN_ALLOW_THREADS/Py_END_ALLOW_THREADS` would be an optimization (in the free-threaded build, it makes threads go into DETACHED state, which makes it a little easier to stop the world for a GC pass), but is not required.

I'm also worried about adding `threadsafe` to more wrapped fortran code, given at least as far as I can tell, it doesn't appear in any of the `pyf` files in the f2py tests.

Reply:
> The other was that existing threadsafe markers are already sufficient to mark a module as compatible if all functions are marked.

I'll take a look at this though, this seems reasonable.

Reply:
So I looked at SciPy, and there is only one module (the `dfitpack` wrappers) where all the functions are marked thread safe. So I don't think it's worth the added complexity in the f2py code to do this, since users can just pass the CLI flag if they want to have free-threading support. I also think manually opting isn't a bad idea either.

Reply:
> since users can just pass the CLI flag if they want to have free-threading support.

Right, but for the next few years at least it would be nice if you could conveniently get GIL release as well.  This way you get free-threaded, I guess, but these should probably already release the GIL?

(have to come back to think about it a bit, maybe it all just doesn't matter much, but it seems like easy to run fully free, but surprsingly hard to release the GIL even though that is roughly the same for fortran code.)

Reply:
I agree in principle, but in practice the f2py wrappers are not written in a way that makes it easy to release the GIL, at least for the code that handles callbacks. I'd also be a little nervous about packages suddenly shipping fortran modules that release the GIL on GIL-enabled builds when they didn't do that before, just because they added support for free-threaded builds. I agree that most of the time it will be fine, but up until now people have had to opt in. I may also be worrying about nothing, it just feels slightly risky to me given the stability of f2py in general.